README.rst
pyblehci
========

pyblehci is a library, written in Python, that provides a combined
parser-builder for TI vendor-specific HCI packets utilised by the
HostTestRelease application.

.. code-block:: pycon

    >>> p = pyblehci.BLEParser()
    >>> p._split_response("\\x04\\xFF\\x08\\x7F\\x06\\x00\\x31\\xFE\\x02\\xD0\\x07")
    ('\\x04\\xff\\x08\\x7f\\x06\\x00\\x31\\xfe\\x02\\xd0\\x07', OrderedDict(
    [('type', ('\\x04', 'Event')),
    ('event_code', ('\\xff', 'HCI_LE_ExtEvent')),
    ('data_len', ('\\x02', '02')),
    ('event', ('\\x06\\x7f', 'GAP_HCI_ExtensionCommandStatus')),
    ('status', ('\\x00', '00')),
    ('op_code', ('\\31\\xfe', 'GAP_GetParam')),
    ('param_value', ('\\xd0\\x07', '07d0'))]))

    >>> b = pyblehci.BLEBuilder()
    >>> b._build_command("fe31", param_id="\x15")
    ('\\x01\\x31\\xfe\\x01\\x15', OrderedDict([
    ('type', ('\\x01', 'Command')),
    ('op_code', ('\\x31\\xfe', 'GAP_GetParam')),
    ('data_len', ('\\x01', '01')),
    ('param_id', ('\\x15', '15'))])
    )
    ...

pyblehci can also be used to read and write to the device found on the serial
interface. The parser operates asynchronously via a callback method allowing
asynchronous reading/writing of the device.
This is possible by use of `pySerial <http://pyserial.sourceforge.net/>`_

.. code-block:: pycon

    >>> serial_port = serial.Serial(port='/dev/ttyACM0', baudrate=57600)
    >>> def analyse_packet(packet):
    ...     print packet
    ...
    >>> p = pyblehci.BLEParser(serial_port, callback=analyse_packet)
    >>> b = pyblehci.BLEBuilder(serial_port)
    >>> b.send(cmd="fe31", param_id="\x15")
    ('\x011\xfe\x01\x15', OrderedDict([('type', ('\x01', 'Command')), ('op_code', ('
    1\xfe', 'GAP_GetParam')), ('data_len', ('\x01', '01')), ('param_id', ('\x15', '1
    5'))]))
    ('\x04\xff\x08\x7f\x06\x001\xfe\x02P\x00', OrderedDict([('type', ('\x04', 'E
    vent')), ('event_code', ('\xff', 'HCI_LE_ExtEvent')), ('data_len', ('\x02', '02'
    )), ('event', ('\x06\x7f', 'GAP_HCI_ExtensionCommandStatus')), ('status', ('\x00
    ', '00')), ('op_code', ('1\xfe', 'GAP_GetParam')), ('param_value', ('P\x00', '00
    50'))]))
    >>>
    ...

Features
--------

- Parsing and Building of TI vendor-specific HCI packets
- Monitoring of serial BLE devices using the HostTestRelease application.

Supported Devices
-----------------

The library should support any TI single-mode BLE device that uses the
HostTestRelease firmware. The below devices are those that have been tested

- TI CC2540DK-MINI Development Kit USB Dongle

Installation
------------

To download pyblehci:

.. code-block:: bash

    $ git clone https://github.com/stephenfin/pyblehci.git

To install:

.. code-block:: bash

    $ python setup.py install

Contribute
----------

- While most of the packets structures have been implemented in both the
  BLEParser and BLEBuilder modules, they are not complete. Additional packets
  can be found in the 'TI Vendor Specific HCI Guide' pdf document that is
  included in the TI BLE stack download

CHANGES.txt
v0.1.0, 05/02/2013 -- Initial release
v0.2.0, 11/02/2013 -- Stable release version
v0.2.1, 22/03/2013 -- Improved documentation
v0.2.2, 22/03/2013 -- Fixed issues caused by previous version documentation
v0.2.3, 07/12/2014 -- Code cleanup, Python 3 fixes


examples/serial-demo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
@fn ble_parser.py

@author Stephen Finucane, 2013-2014
@email  stephenfinucane@hotmail.com

@about Sample application demonstrating use of pyblehci to write/read
    to/from a serial device.
"""

from __future__ import print_function

import collections
import serial
import time

from pyblehci import BLEBuilder
from pyblehci import BLEParser


def pretty(hex_string, seperator=' '):
    """
    Prettify a hex string.

    >>> pretty("\x01\x02\x03\xff")
    '01 02 03 FF'
    """
    hex_string = hex_string.encode('hex')
    out = ''

    for i in range(len(hex_string)):
        if not i % 2:
            out = out + seperator
        out = out + hex_string[i].capitalize()

    return out


def print_ordered_dict(dictionary):
    result = ""
    for key in dictionary:
        if dictionary[key]:
            #convert e.g. "data_len" -> "Data Len"
            title = ' '.join(key.split("_")).title()
            if isinstance(dictionary[key], list):
                for idx2, _ in enumerate(dictionary[key]):
                    result += "{0} ({1})\n".format(title, idx2)
                    result += print_ordered_dict(dictionary[key][idx2])
            elif isinstance(dictionary[key], type(collections.OrderedDict())):
                result += '{0}\n{1}'.format(title, print_ordered_dict(
                    dictionary[key]))
            else:
                result += "{0:15}\t: {1}\n\t\t  ({2})\n".format(
                    title, pretty(dictionary[key][0], ':'), dictionary[key][1])
        else:
            result += "{0:15}\t: None".format(key)
    return result


def print_output((packet, dictionary)):
    result = print_ordered_dict(dictionary)
    result += 'Dump:\n{0}\n'.format(pretty(packet))
    return result


def analyse_packet((packet, dictionary)):
    print("EVENT: Response received from the device")
    print(print_output((packet, dictionary)))


def main():
    """
    Main function.
    """
    # initialise objects
    serial_port = serial.Serial(port='COM4', baudrate=57600)
    ble_builder = BLEBuilder(serial_port)
    ble_parser = BLEParser(serial_port, callback=analyse_packet)

    # initialise the device
    print("COMMAND: Initialising device")
    print(print_output(ble_builder.send("fe00")))

    # get an operating parameter value
    print("COMMAND: Getting operating parameter value")
    print(print_output(ble_builder.send("fe31", param_id="\x15")))

    # start a device discovery scan
    print("COMMAND: Starting device scan")
    print(print_output(ble_builder.send("fe04", mode="\x03")))

    # sleep main thread for 15 seconds - allow results of device scan to return
    time.sleep(15)

    #close device
    ble_parser.stop()


if __name__ == "__main__":
    main()


examples/simple-demo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
@fn ble_parser.py

@author Stephen Finucane, 2013-2014
@email  stephenfinucane@hotmail.com

@about Simple application to build and parse HCI packets from a TI BLE
    device running HostTestRelease.
"""

from __future__ import print_function

from pyblehci import BLEBuilder
from pyblehci import BLEParser

def test_builder():
    ble_builder = BLEBuilder()
    print(ble_builder._build_command("fe00"))
    print(ble_builder._build_command("fe31", param_id="\x15"))
    print(ble_builder._build_command("fe04", mode="\x03"))
    print(ble_builder._build_command(
        "fe09", peer_addr="\x57\x6A\xE4\x31\x18\x00"))
    print(ble_builder._build_command(
        "fd8a", conn_handle="\x00\x00", handle="\x27\x00"))
    print(ble_builder._build_command(
        "fe0a", conn_handle="\x00\x00"))


def test_parser():
    ble_parser = BLEParser()
    print(ble_parser._split_response(
        "\x04\xFF\x2C\x00\x06\x00\x06\x30\x85\x31\x18\x00\x1B\x00\x04\x00\x00\x00\x00\x00\x00\x00\x00\x09\x09\x09\x09\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x09\x09\x09\x09\x00\x00"))
    print(ble_parser._split_response(
        "\x04\xFF\x08\x7F\x06\x00\x31\xFE\x02\xD0\x07"))
    print(ble_parser._split_response(
        "\x04\xFF\x0C\x01\x06\x00\x01\x00\x00\x57\x6A\xE4\x31\x18\x00"))
    print(ble_parser._split_response(
        "\x04\xFF\x14\x01\x06\x00\x01\x00\x00\x57\x6A\xE4\x31\x18\x00\x11\x11\x11\x11\x11\x11\x11\x11"))
    print(ble_parser._split_response(
        "\x04\xFF\x07\x01\x06\x00\x00"))


if __name__ == "__main__":
    test_builder()
    test_parser()


pyblehci/ble_builder.py
"""
@fn ble_builder.py

@author Stephen Finucane, 2013-2014
@email  stephenfinucane@hotmail.com

@about Based heavily on python-xbee library by Paul Malmsten.
    This class defines data and methods applicable to the Texas
    Instruments Bluetooth Low Energy Host-Controller-Interface (HCI)
"""

import collections


class BLEBuilder(object):
    """
    A builder for command packets as defined by the the Texas
    Instruments Bluetooth Low Energy Host-Controller-Interface (HCI).
    """
    # dictionaries
    # opcodes for command packets
    opcodes = {
        "fd8a": 'GATT_ReadCharValue',
        "fd8e": 'GATT_ReadMultipleCharValues',
        "fd92": 'GATT_WriteCharValue',
        "fd96": 'GATT_WriteLongCharValue',
        "fdb2": 'GATT_DiscAllChars',
        "fdb4": 'GATT_ReadUsingCharUUID',
        "fe00": 'GAP_DeviceInit',
        "fe03": 'GAP_ConfigureDeviceAddr',
        "fe04": 'GATT_DeviceDiscoveryRequest',
        "fe05": 'GATT_DeviceDiscoveryCancel',
        "fe09": 'GATT_EstablishLinkRequest',
        "fe0a": 'GATT_TerminateLinkRequest',
        "fe30": 'GAP_SetParam',
        "fe31": 'GAP_GetParam',
    }

    # structure of command packets
    hci_cmds = {
        "fd8a": [
            {'name': 'conn_handle', 'len': 2, 'default': '\x00\x00'},
            {'name': 'handle', 'len': 2, 'default': None}],
        "fd8e": [
            {'name': 'conn_handle', 'len': 2, 'default': '\x00\x00'},
            {'name': 'handles', 'len': None, 'default': None}],
        "fd92": [
            {'name': 'conn_handle', 'len': 2, 'default': '\x00\x00'},
            {'name': 'handle', 'len': 2, 'default': None},
            {'name': 'value', 'len': None, 'default': None}],
        "fd96": [
            {'name': 'handle', 'len': 2, 'default': '\x00\x00'},
            {'name': 'offset', 'len': 1, 'default': None},
            {'name': 'value', 'len': None, 'default': None}],
        "fdb2": [
            {'name': 'start_handle', 'len': 2, 'default': '\x00\x00'},
            {'name': 'end_handle', 'len': 2, 'default': '\xff\xff'}],
        "fdb4": [
            {'name': 'conn_handle', 'len': 2, 'default': '\x00\x00'},
            {'name': 'start_handle', 'len': 2, 'default': '\x01\x00'},
            {'name': 'end_handle', 'len': 2, 'default': '\xff\xff'},
            {'name': 'read_type', 'len': 2, 'default': None}],
        "fe00": [
            {'name': 'profile_role', 'len': 1, 'default': '\x08'},
            {'name': 'max_scan_rsps', 'len': 1, 'default': '\x05'},
            {'name': 'irk', 'len': 16, 'default':
                '\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'},
            {'name': 'csrk', 'len': 16, 'default':
                '\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'},
            {'name': 'sign_counter', 'len': 4, 'default': '\x01\x00\x00\x00'}],
        "fe03": [
            {'name': 'addr_type', 'len': 1, 'default': None},
            {'name': 'addr', 'len': 6, 'default': None}],
        "fe04": [
            {'name': 'mode', 'len': 1, 'default': None},
            {'name': 'active_scan', 'len': 1, 'default': '\x01'},
            {'name': 'white_list', 'len': 1, 'default': '\x00'}],
        "fe05": [],
        "fe09": [
            {'name': 'high_duty_cycle', 'len': 1, 'default': '\x00'},
            {'name': 'white_list', 'len': 1, 'default': '\x00'},
            {'name': 'addr_type_peer', 'len': 1, 'default': '\x00'},
            {'name': 'peer_addr', 'len': 6, 'default': None}],
        "fe0a": [
            {'name': 'conn_handle', 'len': 2, 'default': '\x00\x00'}],
        "fe30": [
            {'name': 'param_id', 'len': 1, 'default': None},
            {'name': 'param_value', 'len': 2, 'default': None}],
        "fe31": [
            {'name': 'param_id', 'len': 1, 'default': None}],
    }

    def __init__(self, ser=None):
        """
        Initialises the class

        @param ser: The file like serial port to use
        @type ser: serial.Serial
        """
        self.serial_port = ser

    def _build_command(self, cmd, **kwargs):
        """
        Constructs a HCI command to the serial port for this BLE
        device.

        >>> _build_command("fe31", param_id="\x15")
        ('\\x01\\x31\\xfe\\x01\\x15', OrderedDict([
        ('type', ('\\x01', 'Command')),
        ('op_code', ('\\x31\\xfe', 'GAP_GetParam')),
        ('data_len', ('\\x01', '01')),
        ('param_id', ('\\x15', '15'))])
        )

        It will expect named arguments for all fields other than those
        with a default value or length of 'None'.

        >>> _build_command("fe31")
        Traceback (most recent call last):
        File "<stdin>", line 1, in <module>
        File "C:\Python27\lib\site-packages\pyblehci\\ble_builder.py", line \
168, in _build_command % (field_name, field_len))
        KeyError: "The data provided for 'param_id' was not 1 bytes long"

        Each field will be written out in the order they are defined in
        the command definition.

        @param cmd: The command to be written
        @type cmd: hex

        @param kwargs: Any additional parameters
        @type kwargs: hex

        @return: A tuple containing the hex command string and a parsed
            version of the string stored in a dictionary.
        """
        packet_type = "\x01"
        op_code = cmd.decode('hex')[::-1]  # command code was human-readable
        data_len = "\x00"  # insert dummy value for length

        # check for matching command codes in dictionary and store the matching
        # packet format
        try:
            packet_structure = self.hci_cmds[cmd]
        except AttributeError:
            raise NotImplementedError("Command spec could not be found")

        packet_type_parsed = "Command"
        op_code_parsed = self.opcodes[cmd]
        data_len_parsed = "0"  # insert dummy value for length

        # command match found, hence start storing result
        built_packet = collections.OrderedDict()
        built_packet['type'] = (packet_type, packet_type_parsed)
        built_packet['op_code'] = (op_code, op_code_parsed)
        built_packet['data_len'] = (data_len, data_len_parsed)

        packet = ''
        packet += packet_type
        packet += op_code
        packet += data_len

        # build the packet in the order specified, by processing each
        # required value as needed
        for field in packet_structure:
            field_name = field['name']
            field_len = field['len']
            # try to read this field's name from the function arguments dict
            try:
                field_data = kwargs[field_name]
            # data wasn't given
            except KeyError:
                # only a problem is the field has a specific length...
                if field_len is not None:
                    #...or a default value
                    default_value = field['default']
                    if default_value:
                        field_data = default_value
                    else:
                        raise KeyError(
                            "The data provided for '%s' was not %d bytes long"
                            % (field_name, field_len))
                # no specific length, hence ignore it
                else:
                    field_data = None

            # ensure that the correct number of elements will be written
            if field_len and len(field_data) != field_len:
                raise ValueError(
                    "The data provided for '%s' was not %d bytes long"
                    % (field_name, field_len))

            # add the data to the packet if it has been specified (otherwise
            # the parameter was of variable length and not given)
            if field_data:
                packet += field_data
                built_packet[field_name] = (
                    field_data, field_data.encode('hex'))

        # finally, replace the dummy length value in the string
        length = hex(len(packet) - 4)  # get length of bytes after 4th (length)
        data_len = length[2:].zfill(2).decode('hex')  # change 0x2 -> \x02
        modified_packet = list(packet)
        modified_packet[3] = data_len
        packet = "".join(modified_packet)
        # and the dictionary
        data_len_parsed = data_len.encode('hex')
        built_packet['data_len'] = (data_len, data_len_parsed)

        return (packet, built_packet)

    def send(self, cmd, **kwargs):
        """
        Constructs and write a HCI command to the serial port for this
        BLE device.

        >>> send(cmd="fe31", param_id="\x15")
        01:31:FE:01:15  #<-- also writes this to serial port

        This method must be called with the named arguments in accordance
        with the HCI specification. Arguments matching all field names
        other than those in the reserved_names (like 'id' and 'order')
        should be given, unless they are of variable length (of 'None' in
        the specification. These are optional).

        >>> send(cmd="fe31")
        Traceback (most recent call last):
        File "<stdin>", line 1, in <module>
        File "C:\Python27\lib\site-packages\pyblehci\\ble_builder.py", line 168, \
in _build_command % (field_name, field_len))
        KeyError: "The data provided for 'param_id' was not 1 bytes long"

        Each field will be written out in the order they are defined in
        the command definition.

        @param cmd: The command to be written
        @type cmd: hex

        @param kwargs: Any additional parameters
        @type kwargs: hex

        @return: A tuple containing the hex command string and a parsed
        version of the string stored in a dictionary.
        """
        packet, built_packet = self._build_command(cmd, **kwargs)
        self.serial_port.write(packet)

        return (packet, built_packet)


pyblehci/ble_parser.py
"""
@fn ble_parser.py

@author Stephen Finucane, 2013-2014
@email  stephenfinucane@hotmail.com

@about Based heavily on python-xbee library by Paul Malmsten.
    This class defines data and methods applicable to the Texas
    Instruments Bluetooth Low Energy Host-Controller-Interface (HCI)
"""

import collections
import threading
import time


class ThreadQuitException(Exception):
    """
    Override thread quit exception handler.
    """
    pass


class BLEParser(threading.Thread):
    """
    A parser for event packets as defined by the the Texas Instruments
    Bluetooth Low Energy Host-Controller-Interface (HCI).

    Capable of monitoring a serial device, parsing the packets and
    returning to a calling method by callback.
    """
    # dictionaries
    # opcodes for command packets
    opcodes = {
        "fd8a": 'GATT_ReadCharValue',
        "fd8e": 'GATT_ReadMultipleCharValues',
        "fd92": 'GATT_WriteCharValue',
        "fd96": 'GATT_WriteLongCharValue',
        "fdb2": 'GATT_DiscAllChars',
        "fdb4": 'GATT_ReadUsingCharUUID',
        "fe00": 'GAP_DeviceInit',
        "fe03": 'GAP_ConfigureDeviceAddr',
        "fe04": 'GATT_DeviceDiscoveryRequest',
        "fe05": 'GATT_DeviceDiscoveryCancel',
        "fe09": 'GATT_EstablishLinkRequest',
        "fe0a": 'GATT_TerminateLinkRequest',
        "fe30": 'GAP_SetParam',
        "fe31": 'GAP_GetParam',
    }

    # structure of event packets
    hci_events = {
        "ff": {
            'name': 'HCI_LE_ExtEvent',
            'structure': [{'name': 'ext_event', 'len': None}]},
        }

    # parameter formats for HCI_LE_ExtEvent
    ext_events = {
        "0501": {
            'name': 'ATT_ErrorRsp',
            'structure': [
                {'name': 'conn_handle', 'len': 2},
                {'name': 'pdu_len', 'len': 1},
                {'name': 'req_op_code', 'len': 1},
                {'name': 'handle', 'len': 2},
                {'name': 'error_code', 'len': 1}]},
        "0509": {
            'name': 'ATT_ReadByTypeRsp',
            'structure': [
                {'name': 'conn_handle', 'len': 2},
                {'name': 'pdu_len', 'len': 1},
                {'name': 'length', 'len': 1},
                {'name': 'results', 'len': None}],
            'parsing': [
                ('results', lambda ble, original:
                 ble._parse_read_results(original['results']))]},
        "050b": {
            'name': 'ATT_ReadRsp',
            'structure': [
                {'name': 'conn_handle', 'len': 2},
                {'name': 'pdu_len', 'len': 1},
                {'name': 'value', 'len': None}]},
        "050f": {
            'name': 'ATT_ReadMultiRsp',
            'structure': [
                {'name': 'conn_handle', 'len': 2},
                {'name': 'pdu_len', 'len': 1},
                {'name': 'results', 'len': None}]},
        "0513": {
            'name': 'ATT_WriteRsp',
            'structure': [
                {'name': 'conn_handle', 'len': 2},
                {'name': 'pdu_len', 'len': 1}]},
        "051b": {
            'name': 'ATT_HandleValueNotification',
            'structure': [
                {'name': 'conn_handle', 'len': 2},
                {'name': 'pdu_len', 'len': 1},
                {'name': 'handle', 'len': 2},
                {'name': 'values', 'len': None}]},
        "0600": {
            'name': 'GAP_DeviceInitDone',
            'structure': [
                {'name': 'dev_addr', 'len': 6},
                {'name': 'data_pkt_len', 'len': 2},
                {'name': 'num_data_pkts', 'len': 1},
                {'name': 'irk', 'len': 16},
                {'name': 'csrk', 'len': 16}]},
        "0601": {
            'name': 'GAP_DeviceDiscoveryDone',
            'structure': [
                {'name': 'num_devs', 'len': 1},
                {'name': 'devices', 'len': None}],
            'parsing': [
                ('devices', lambda ble, original:
                 ble._parse_devices(original['devices']))]},
        "0605": {
            'name': 'GAP_EstablishLink',
            'structure': [
                {'name': 'dev_addr_type', 'len': 1},
                {'name': 'dev_addr', 'len': 6},
                {'name': 'conn_handle', 'len': 2},
                {'name': 'conn_interval', 'len': 2},
                {'name': 'conn_latency', 'len': 2},
                {'name': 'conn_timeout', 'len': 2},
                {'name': 'clock_accuracy', 'len': 1}]},
        "060d": {
            'name': 'GAP_DeviceInformation',
            'structure': [
                {'name': 'event_type', 'len': 1},
                {'name': 'addr_type', 'len': 1},
                {'name': 'addr', 'len': 6},
                {'name': 'rssi', 'len': 1},
                {'name': 'data_len', 'len': 1},
                {'name': 'data_field', 'len': None}]},
        "0606": {
            'name': 'GAP_LinkTerminated',
            'structure': [
                {'name': 'conn_handle', 'len': 2},
                {'name': 'reason', 'len': 1}]},
        "067f": {
            'name': 'GAP_HCI_ExtensionCommandStatus',
            'structure': [
                {'name': 'op_code', 'len': 2},
                {'name': 'data_len', 'len': 1},
                {'name': 'param_value', 'len': None}],
            'parsing': [
                ('op_code', lambda ble, original:
                 ble._parse_opcodes(original['op_code']))]},
    }

    def __init__(self, ser=None, callback=None):
        """
        Initialises the class

        @param ser: The file like serial port to use
        @type ser: serial.Serial

        @param callback: The callback method
        @type callback: <function>
        """
        super(BLEParser, self).__init__()
        self.serial_port = ser
        self._callback = None
        self._thread_continue = False
        self._stop = threading.Event()

        if callback:
            self._callback = callback
            self._thread_continue = True
            self.start()

    def run(self):
        """
        Overrides threading.Thread.run() and is automatically
        called when an instance is created with threading enabled
        """
        while True:
            try:
                self._callback(self.wait_read())
            except ThreadQuitException:
                break

    def stop(self):
        """
        Stops the thread and closes the serial port
        """
        self._thread_continue = False
        self.serial_port.close()
        self._stop.set()

    def stopped(self):
        """
        Getter method for isSet variable

        >>> stopped()
        false
        """
        return self._stop.isSet()

    def _wait_for_frame(self):
        """
        Reads from the serial port until a valid HCI packet arrives. It
        will then return the binary data contained within the packet.

        @return: A byte string of the correct length
        """
        # loop forever...
        while True:
            #...unless told not to by setting "_thread_continue" to false
            if self._callback and not self._thread_continue:
                raise ThreadQuitException

            # prevent blocking the port by waiting a given time
            # TODO. Remove this? Asynchronous read-write possible
            if self.serial_port.inWaiting() == 0:
                time.sleep(.01)
                continue

            # length byte is stored as the third byte in an event packet
            packet = self.serial_port.read(3)
            # convert this to decimal...
            data_len = int(packet[2].encode('hex'), 16)

            #...and retrieve that many bytes from the serial port
            for _ in range(0, data_len):
                packet += self.serial_port.read()

            return packet

    def _split_response(self, data):
        """
        Takes a data packet received from a TI BLE device and parses
        it.

        >>> _split_response("\\x04\\xFF\\x08\\x7F\\x06\\x00\\x31\\xFE\\x02\\xD0\\x07")
        ('\\x04\\xff\\x08\\x7f\\x06\\x00\\x31\\xfe\\x02\\xd0\\x07', OrderedDict(
        [('type', ('\\x04', 'Event')),
        ('event_code', ('\\xff', 'HCI_LE_ExtEvent')),
        ('data_len', ('\\x02', '02')),
        ('event', ('\\x06\\x7f', 'GAP_HCI_ExtensionCommandStatus')),
        ('status', ('\\x00', '00')),
        ('op_code', ('\\31\\xfe', 'GAP_GetParam')),
        ('param_value', ('\\xd0\\x07', '07d0'))]))

        @param data: The byte string to split and parse
        @type data: hex

        @return: An ordered dictionary (data order is important)
            containing binary tuples, in which the first piece of data
            corresponds to the raw byte string value and the second
            piece corresponds to its parsed "meaning"
        """
        packet_type = data[0]
        event_code = data[1]
        data_len = data[2]

        # check for matching event codes in dictionary and store the matching
        # packet format
        try:
            packet = self.hci_events[event_code.encode('hex')]
        except AttributeError:
            raise NotImplementedError("Error with Attribute")
        except KeyError:
            raise KeyError("Unrecognized response packet with event" +
                           " type {0}".format(event_code.encode('hex')))

        packet_type_parsed = "Event"
        event_code_parsed = packet['name']
        data_len_parsed = int(data_len.encode('hex'), 16)

        # packet match found, hence start storing result
        parsed_packet = collections.OrderedDict()
        parsed_packet['type'] = (packet_type, packet_type_parsed)
        parsed_packet['event_code'] = (event_code, event_code_parsed)
        parsed_packet['data_len'] = (data_len, data_len_parsed)

        # special handler for HCI_LE_ExtEvent
        if event_code_parsed == 'HCI_LE_ExtEvent':
            # current byte index in the data stream
            index = 6
            # event_subcode is two-bytes given in reverse (endian mismatch?)
            event_subcode = data[3:5][::-1]  # reverse byte order [::-1]
            event_status = data[5]

            try:
                subpacket = self.ext_events[event_subcode.encode('hex')]
            except AttributeError:
                raise NotImplementedError("Error with Attribute")
            except KeyError:
                print data.encode('hex')
                raise KeyError("Unrecognized response packet with event" +
                               " type {0}".format(data[3:5][::-1]))

            event_subcode_parsed = subpacket['name']
            event_status_parsed = event_status.encode('hex')

            # subpacket match found, hence store result
            parsed_packet['event'] = (event_subcode, event_subcode_parsed)
            parsed_packet['status'] = (event_status, event_status_parsed)

            # store the subpacket structure for working with in next step
            subpacket_structure = subpacket['structure']

            # parse the subpacket in the order specified, by processing each
            # required value as needed
            for field in subpacket_structure:
                field_name = field['name']
                # if the data field has a fixed length, process it normally
                if field['len'] is not None:
                    # store the number of bytes specified in the dictionary
                    field_data = data[index:(index + field['len'])]
                    field_data_parsed = field_data[::-1].encode('hex')
                    # store result
                    parsed_packet[field_name] = (field_data, field_data_parsed)
                    # increment index for next field
                    index += field['len']
                # if the data field has no length specified, store any leftover
                # bytes and quit
                else:
                    field_data = data[index:]
                    # were there any remaining bytes?
                    if field_data:
                        # if so, store them
                        field_data_parsed = field_data[::-1].encode('hex')
                        parsed_packet[field_name] = (
                            field_data, field_data_parsed)
                        index += len(field_data)
                    break

            # check if there are remaining bytes. If so, raise an exception
            if index < data_len_parsed:
                raise ValueError("Response packet was longer than expected;" +
                                 "expected: %d, got: %d bytes" % (index, data_len_parsed))

            # check for parsing rules and apply them if they exist
            if 'parsing' in subpacket:
                for parse_rule in subpacket['parsing']:
                    # only apply a rule if relevant (raw data available)
                    parse_rule_name = parse_rule[0]
                    parse_rule_def = parse_rule[1]
                    if parse_rule_name in parsed_packet:
                        # apply the parse function to the indicated field
                        # and replace the raw data with the result
                        parsed_packet[parse_rule_name] = parse_rule_def(
                            self, parsed_packet)

        return (data, parsed_packet)

    def _parse_opcodes(self, parsed_packet):
        """
        Functions as a special parsing routine for the "GAP HCI
        Extention Command Status" HCI LE ExtEvent.

        >>> _parse_opcodes(("\\x04\\xFE", "fe04"))
        ("\\x04\\xFE", "GAP_DeviceDiscoveryRequest")

        @param parsed_packet: A tuple of a byte string and the ascii
            encoded copy
        @type parsed_packet: (hex, string)

        @return: An ordered dictionary (data order is important)
            containing binary tuples, in which the first piece of data
            corresponds to the raw byte string value and the second
            piece corresponds to its parsed "meaning" - the command
            name sourced by lookup of the command dict
        """
        value = self.opcodes[parsed_packet[1]]
        return (parsed_packet[0], value)

    def _parse_devices(self, orig_devices):
        """
        Functions as a special parsing routine for the "GAP Device
        Discovery Done" HCI LE ExtEvent.

        >>> _parse_devices(("\\x00\\x00\\x57\\x6A\\xE4\\x31\\x18\\x00", "0000576AE4311800"))
        [OrderedDict([
        ('event_type', ('\\x00', '00')),
        ('addr_type', ('\\x00', '00')),
        ('addr', ('\\x57\\x6a\\xe4\\x31\\x18\\x00', '001831e46a57'))
        ])]

        @param orig_devices: A tuple of a byte string and the ascii
            encoded copy
        @type orig_devices: (hex, string)

        @return: An ordered dictionary (data order is important)
            containing binary tuples, in which the first piece of data
            corresponds to the raw byte string value and the second
            piece corresponds to its parsed "meaning" - currently just
            the hex encoded version of the string
        """
        parsed_devices = []
        # seperate the byte string containing the devices into groups of eight
        # bytes
        for device in [orig_devices[0][i:i + 8] for i in
                       range(0, len(orig_devices[0]), 8)]:
            event_type = device[0]
            addr_type = device[1]
            addr = device[2:9]

            event_type_parsed = event_type.encode('hex')
            addr_type_parsed = addr_type.encode('hex')
            addr_parsed = addr[::-1].encode('hex')

            # store the parsed device as an ordered dictionary (order once again
            # important)
            temp_device = collections.OrderedDict()
            temp_device['event_type'] = (event_type, event_type_parsed)
            temp_device['addr_type'] = (addr_type, addr_type_parsed)
            temp_device['addr'] = (addr, addr_parsed)

            # append the ordered dict containing the parsed device to a list
            parsed_devices.append(temp_device)
        # return the resulting list
        return parsed_devices

    def _parse_read_results(self, results):
        """
        Functions as a special parsing routine for the "ATT Read By
        Type Rsp" HCI LE ExtEvent.

        >>> _parse_read_results(("\\x00\\x00\\x57\\x6A\\xE4\\x31\\x18\\x00", "0000576AE4311800"))
        TODO

        @param results: A tuple of a byte string and the ascii encoded
            copy
        @type results: (hex, string)

        @return: An ordered dictionary (data order is important)
            containing binary tuples, in which the first piece of data
            corresponds to the raw byte string value and the second
            piece corresponds to its parsed "meaning" - currently just
            the hex encoded version of the string
        """
        parsed_results = []
        # seperate the byte string containing the results into groups of eight
        # bytes
        for result in [results[0][i:i + 8] for i in
                       range(0, len(results[0]), 8)]:
            handle = result[0:2]
            data = result[2:9]

            handle_parsed = handle[::-1].encode('hex')
            data_parsed = data[::-1].encode('hex')

            # store the parsed result as an ordered dictionary (order once again
            # important)
            temp_result = collections.OrderedDict()
            temp_result['handle'] = (handle, handle_parsed)
            temp_result['data'] = (data, data_parsed)

            # append the ordered dict containing the parsed result to a list
            parsed_results.append(temp_result)
        # return the resulting list
        return parsed_results

    def wait_read(self):
        """
        Combines both '_wait_for_frame' (to read a valid packet) and
        '_split_response' (to parse that packet).

        @return: A parsed version of the packet received on the serial
            port
        """
        packet = self._wait_for_frame()
        return self._split_response(packet)
